File: Sources/KalshiQuickSearchApp/Services/KalshiAPIClient.swift
import Foundation

struct KalshiOpenEventsResponse: Decodable {
    let events: [KalshiEventDTO]
    let cursor: String?
}

struct KalshiEventDTO: Decodable {
    let eventTicker: String?
    let seriesTicker: String?
    let title: String?
    let subtitle: String?
    let category: String?
    let markets: [KalshiMarketDTO]?
    let productMetadata: [String: JSONValue]?
    let url: String?
    let slug: String?

    enum CodingKeys: String, CodingKey {
        case eventTicker = "event_ticker"
        case seriesTicker = "series_ticker"
        case title
        case subtitle = "sub_title"
        case category
        case markets
        case productMetadata = "product_metadata"
        case url
        case slug
    }
}

struct KalshiMarketDTO: Decodable {
    let ticker: String
    let title: String?
    let subtitle: String?
    let yesLabel: String?
    let noLabel: String?
    let status: String?
    let lastPrice: Double?
    let yesPrice: Double?
    let noPrice: Double?
    let volume: Double?
    let volume24h: Double?
    let openInterest: Double?
    let liquidity: Double?
    let updatedTime: Date?
    let openTime: Date?
    let closeTime: Date?
    let rulesPrimary: String?
    let rulesSecondary: String?
    let url: String?
    let slug: String?
    let result: String?
    let productMetadata: [String: JSONValue]?

    enum CodingKeys: String, CodingKey {
        case ticker
        case title
        case subtitle = "sub_title"
        case yesLabel = "yes_sub_title"
        case noLabel = "no_sub_title"
        case status
        case lastPrice = "last_price_dollars"
        case yesPrice = "yes_ask_dollars"
        case noPrice = "no_ask_dollars"
        case volume = "volume_fp"
        case volume24h = "volume_24h_fp"
        case openInterest = "open_interest_fp"
        case liquidity = "liquidity_dollars"
        case updatedTime = "updated_time"
        case openTime = "open_time"
        case closeTime = "close_time"
        case rulesPrimary = "rules_primary"
        case rulesSecondary = "rules_secondary"
        case url
        case slug
        case result
        case productMetadata = "product_metadata"
        case yesBid = "yes_bid_dollars"
        case noBid = "no_bid_dollars"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        ticker = try container.decode(String.self, forKey: .ticker)
        title = try container.decodeIfPresent(String.self, forKey: .title)
        subtitle = try container.decodeIfPresent(String.self, forKey: .subtitle)
        yesLabel = try container.decodeIfPresent(String.self, forKey: .yesLabel)
        noLabel = try container.decodeIfPresent(String.self, forKey: .noLabel)
        status = try container.decodeIfPresent(String.self, forKey: .status)

        let last = container.decodeFlexibleDouble(forKey: .lastPrice)
        let yesAsk = container.decodeFlexibleDouble(forKey: .yesPrice)
        let noAsk = container.decodeFlexibleDouble(forKey: .noPrice)
        let yesBid = container.decodeFlexibleDouble(forKey: .yesBid)
        let noBid = container.decodeFlexibleDouble(forKey: .noBid)

        lastPrice = last
        yesPrice = yesAsk ?? yesBid ?? last
        noPrice = noAsk ?? noBid ?? last.map { max(0, 1 - $0) }

        volume = container.decodeFlexibleDouble(forKey: .volume)
        volume24h = container.decodeFlexibleDouble(forKey: .volume24h)
        openInterest = container.decodeFlexibleDouble(forKey: .openInterest)
        liquidity = container.decodeFlexibleDouble(forKey: .liquidity)
        updatedTime = container.decodeFlexibleDate(forKey: .updatedTime)
        openTime = container.decodeFlexibleDate(forKey: .openTime)
        closeTime = container.decodeFlexibleDate(forKey: .closeTime)
        rulesPrimary = try container.decodeIfPresent(String.self, forKey: .rulesPrimary)
        rulesSecondary = try container.decodeIfPresent(String.self, forKey: .rulesSecondary)
        url = try container.decodeIfPresent(String.self, forKey: .url)
        slug = try container.decodeIfPresent(String.self, forKey: .slug)
        result = try container.decodeIfPresent(String.self, forKey: .result)
        productMetadata = try container.decodeIfPresent([String: JSONValue].self, forKey: .productMetadata)
    }
}

struct KalshiBatchCandlestickResponse: Decodable {
    let candlesticks: [KalshiMarketCandlestickSeries]
}

struct KalshiMarketCandlestickSeries: Decodable {
    let marketTicker: String
    let series: [KalshiCandleDTO]

    enum CodingKeys: String, CodingKey {
        case marketTicker = "market_ticker"
        case series = "candlesticks"
    }
}

struct KalshiCandleDTO: Decodable {
    let endTs: Int64?
    let closePrice: Double?
    let yesPrice: Double?

    enum CodingKeys: String, CodingKey {
        case endTs = "end_period_ts"
        case closePrice = "close"
        case yesPrice = "yes_price"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        endTs = container.decodeFlexibleInt64(forKey: .endTs)
        closePrice = container.decodeFlexibleDouble(forKey: .closePrice)
        yesPrice = container.decodeFlexibleDouble(forKey: .yesPrice)
    }
}

struct KalshiAPIClient: Sendable {
    private let session: URLSession
    private let decoder: JSONDecoder
    private let baseURL = URL(string: "https://api.elections.kalshi.com/trade-api/v2")!

    init() {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 20
        configuration.timeoutIntervalForResource = 30
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        self.session = URLSession(configuration: configuration)
        self.decoder = JSONDecoder()
    }

    func fetchOpenMarkets() async throws -> [MarketSummary] {
        var cursor: String?
        var allMarkets: [MarketSummary] = []

        repeat {
            var components = URLComponents(url: baseURL.appendingPathComponent("events"), resolvingAgainstBaseURL: false)!
            var queryItems = [
                URLQueryItem(name: "status", value: "open"),
                URLQueryItem(name: "limit", value: "200"),
                URLQueryItem(name: "with_nested_markets", value: "true")
            ]
            if let cursor {
                queryItems.append(URLQueryItem(name: "cursor", value: cursor))
            }
            components.queryItems = queryItems

            let response: KalshiOpenEventsResponse = try await get(url: components.url!)
            let markets = response.events.flatMap { event in
                (event.markets ?? [])
                    .filter { market in
                        let normalized = (market.status ?? "").lowercased()
                        return normalized == "active" || normalized == "open"
                    }
                    .map { $0.toDomain(event: event) }
            }
            allMarkets.append(contentsOf: markets)
            cursor = response.cursor
        } while cursor != nil

        return allMarkets
    }

    func fetchTrend(for marketTicker: String, window: TrendWindow) async throws -> MarketTrend {
        let endDate = Date()
        let startDate = endDate.addingTimeInterval(-window.duration)
        var components = URLComponents(url: baseURL.appendingPathComponent("markets/candlesticks"), resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "market_tickers", value: marketTicker),
            URLQueryItem(name: "start_ts", value: String(Int64(startDate.timeIntervalSince1970))),
            URLQueryItem(name: "end_ts", value: String(Int64(endDate.timeIntervalSince1970))),
            URLQueryItem(name: "period_interval", value: String(window.intervalMinutes)),
            URLQueryItem(name: "include_latest_before_start", value: "true")
        ]

        let response: KalshiBatchCandlestickResponse = try await get(url: components.url!)
        let series = response.candlesticks.first(where: { $0.marketTicker == marketTicker })

        let points = (series?.series ?? []).compactMap { candle -> TrendPoint? in
            let timestamp = candle.endTs
            let rawPrice = candle.closePrice ?? candle.yesPrice
            guard let timestamp, let rawPrice else { return nil }
            let normalizedPrice = rawPrice > 1 ? rawPrice : rawPrice * 100
            return TrendPoint(date: Date(timeIntervalSince1970: TimeInterval(timestamp)), value: normalizedPrice)
        }.sorted { $0.date < $1.date }

        let delta: Double?
        if let first = points.first?.value, let last = points.last?.value {
            delta = last - first
        } else {
            delta = nil
        }

        return MarketTrend(
            marketTicker: marketTicker,
            window: window,
            points: points,
            delta: delta,
            updatedAt: Date()
        )
    }

    private func get<T: Decodable>(url: URL) async throws -> T {
        var request = URLRequest(url: url)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse, 200..<300 ~= httpResponse.statusCode else {
            throw URLError(.badServerResponse)
        }

        return try decoder.decode(T.self, from: data)
    }
}

private extension KalshiMarketDTO {
    func toDomain(event: KalshiEventDTO) -> MarketSummary {
        let combinedMetadata = [event.productMetadata, productMetadata]
            .compactMap { $0 }
            .reduce(into: [String: JSONValue]()) { partial, next in
                partial.merge(next) { current, _ in current }
            }

        let imageURL = (combinedMetadata["image"]?.stringValue)
            .flatMap(URL.init(string:))
            ?? combinedMetadata.values.flatMap { $0.collectURLStrings() }
                .compactMap(URL.init(string:))
                .first(where: { $0.absoluteString.lowercased().contains("image") || $0.pathExtension.lowercased().matchesAny(of: ["png", "jpg", "jpeg", "webp"]) })

        let url = [self.url, event.url]
            .compactMap { $0 }
            .compactMap(URL.init(string:))
            .first

        let description = [rulesPrimary, rulesSecondary].compactMap { $0 }.joined(separator: "\n\n").nilIfEmpty

        return MarketSummary(
            ticker: ticker,
            eventTicker: event.eventTicker,
            seriesTicker: event.seriesTicker,
            title: title ?? event.title ?? ticker,
            subtitle: subtitle,
            yesLabel: yesLabel,
            noLabel: noLabel,
            eventTitle: event.title,
            eventSubtitle: event.subtitle,
            description: description,
            category: event.category,
            status: status ?? "active",
            lastPrice: normalizeUnitPrice(lastPrice),
            yesPrice: normalizeUnitPrice(yesPrice),
            noPrice: normalizeUnitPrice(noPrice),
            volume: volume,
            volume24h: volume24h,
            openInterest: openInterest,
            liquidity: liquidity,
            updatedAt: updatedTime,
            openTime: openTime,
            closeTime: closeTime,
            imageURL: imageURL,
            webURL: url
        )
    }

    private func normalizeUnitPrice(_ value: Double?) -> Double? {
        guard let value else { return nil }
        return value > 1 ? value / 100 : value
    }
}

private extension KeyedDecodingContainer {
    func decodeFlexibleDouble(forKey key: Key) -> Double? {
        if let double = try? decodeIfPresent(Double.self, forKey: key) {
            return double
        }
        if let int = try? decodeIfPresent(Int.self, forKey: key) {
            return Double(int)
        }
        if let string = try? decodeIfPresent(String.self, forKey: key) {
            return Double(string)
        }
        return nil
    }

    func decodeFlexibleInt64(forKey key: Key) -> Int64? {
        if let int = try? decodeIfPresent(Int64.self, forKey: key) {
            return int
        }
        if let int = try? decodeIfPresent(Int.self, forKey: key) {
            return Int64(int)
        }
        if let string = try? decodeIfPresent(String.self, forKey: key) {
            return Int64(string)
        }
        return nil
    }

    func decodeFlexibleDate(forKey key: Key) -> Date? {
        if let date = try? decodeIfPresent(Date.self, forKey: key) {
            return date
        }
        guard let string = (try? decodeIfPresent(String.self, forKey: key)) ?? nil else {
            return nil
        }
        return Date.parsedKalshiTimestamp(string)
    }
}

private extension Date {
    static func parsedKalshiTimestamp(_ string: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = fractional.date(from: string) {
            return date
        }
        let basic = ISO8601DateFormatter()
        basic.formatOptions = [.withInternetDateTime]
        return basic.date(from: string)
    }
}

private extension String {
    func matchesAny(of extensions: [String]) -> Bool {
        extensions.contains(self)
    }

    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}

File: packages/android/app/src/main/java/com/truthpulse/data/KalshiApiClient.kt
package com.truthpulse.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.*
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

class KalshiApiClient {

    private val baseUrl = "https://api.elections.kalshi.com/trade-api/v2"

    private val client = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        coerceInputValues = true
    }

    /**
     * Fetch all open markets with cursor pagination.
     */
    suspend fun fetchOpenMarkets(): List<MarketSummary> = withContext(Dispatchers.IO) {
        val allMarkets = mutableListOf<MarketSummary>()
        var cursor: String? = null

        do {
            val url = buildString {
                append("$baseUrl/events?status=open&limit=200&with_nested_markets=true")
                if (!cursor.isNullOrEmpty()) {
                    append("&cursor=$cursor")
                }
            }

            val request = Request.Builder()
                .url(url)
                .header("Accept", "application/json")
                .get()
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                response.close()
                throw RuntimeException("HTTP ${response.code}: ${response.message}")
            }

            val body = response.body?.string() ?: throw RuntimeException("Empty response body")
            response.close()

            val root = json.parseToJsonElement(body).jsonObject
            val events = root["events"]?.jsonArray ?: break
            if (events.isEmpty()) break

            for (eventElement in events) {
                val event = eventElement.jsonObject
                val markets = event["markets"]?.jsonArray ?: continue

                for (marketElement in markets) {
                    val market = marketElement.jsonObject
                    val status = market.stringOrNull("status")?.lowercase() ?: ""
                    if (status != "active" && status != "open") continue

                    allMarkets.add(parseMarket(market, event))
                }
            }

            cursor = root.stringOrNull("cursor")
        } while (!cursor.isNullOrEmpty())

        allMarkets
    }

    /**
     * Fetch candlestick trend data for a specific market.
     */
    suspend fun fetchTrend(
        marketTicker: String,
        seriesTicker: String,
        window: TrendWindow,
    ): MarketTrend = withContext(Dispatchers.IO) {
        val now = System.currentTimeMillis() / 1000
        val startTs = now - window.durationSeconds

        val url = "$baseUrl/series/$seriesTicker/markets/$marketTicker/candlesticks" +
            "?start_ts=$startTs&end_ts=$now&period_interval=${window.intervalMinutes}"

        val request = Request.Builder()
            .url(url)
            .header("Accept", "application/json")
            .get()
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            response.close()
            throw RuntimeException("HTTP ${response.code}: ${response.message}")
        }

        val body = response.body?.string() ?: throw RuntimeException("Empty response body")
        response.close()

        val root = json.parseToJsonElement(body).jsonObject
        val candlesticks = root["candlesticks"]?.jsonArray ?: JsonArray(emptyList())

        val points = candlesticks.mapNotNull { element ->
            val candle = element.jsonObject
            val timestamp = candle.flexibleLong("end_period_ts") ?: return@mapNotNull null
            val rawPrice = bestCandlePrice(candle) ?: return@mapNotNull null
            val normalizedPrice = if (rawPrice > 1.0) rawPrice else rawPrice * 100.0
            TrendPoint(
                timestampMillis = timestamp * 1000,
                value = normalizedPrice,
            )
        }.sortedBy { it.timestampMillis }

        val delta = if (points.size >= 2) {
            points.last().value - points.first().value
        } else null

        MarketTrend(
            marketTicker = marketTicker,
            window = window,
            points = points,
            delta = delta,
            updatedAtMillis = System.currentTimeMillis(),
        )
    }

    private fun bestCandlePrice(candle: JsonObject): Double? {
        // yes_ask.close_dollars -> yes_bid.close_dollars -> price.previous_dollars
        candle["yes_ask"]?.jsonObject?.flexibleDouble("close_dollars")?.let { return it }
        candle["yes_bid"]?.jsonObject?.flexibleDouble("close_dollars")?.let { return it }
        candle["price"]?.jsonObject?.flexibleDouble("previous_dollars")?.let { return it }
        return null
    }

    private fun parseMarket(market: JsonObject, event: JsonObject): MarketSummary {
        val lastPrice = market.flexibleDouble("last_price_dollars")
        val yesAsk = market.flexibleDouble("yes_ask_dollars")
        val noAsk = market.flexibleDouble("no_ask_dollars")
        val yesBid = market.flexibleDouble("yes_bid_dollars")
        val noBid = market.flexibleDouble("no_bid_dollars")

        val yesPrice = yesAsk ?: yesBid ?: lastPrice
        val noPrice = noAsk ?: noBid ?: lastPrice?.let { maxOf(0.0, 1.0 - it) }

        val imageUrl = extractImageUrl(event, market)
        val webUrl = market.stringOrNull("url") ?: event.stringOrNull("url")

        return MarketSummary(
            ticker = market.stringOrNull("ticker") ?: "",
            eventTicker = event.stringOrNull("event_ticker"),
            seriesTicker = event.stringOrNull("series_ticker"),
            title = market.stringOrNull("title") ?: event.stringOrNull("title") ?: "",
            subtitle = market.stringOrNull("sub_title"),
            yesLabel = market.stringOrNull("yes_sub_title"),
            noLabel = market.stringOrNull("no_sub_title"),
            eventTitle = event.stringOrNull("title"),
            eventSubtitle = event.stringOrNull("sub_title"),
            category = event.stringOrNull("category"),
            status = market.stringOrNull("status") ?: "active",
            lastPrice = normalizeUnitPrice(lastPrice),
            yesPrice = normalizeUnitPrice(yesPrice),
            noPrice = normalizeUnitPrice(noPrice),
            volume = market.flexibleDouble("volume_fp"),
            volume24h = market.flexibleDouble("volume_24h_fp"),
            openInterest = market.flexibleDouble("open_interest_fp"),
            liquidity = market.flexibleDouble("liquidity_dollars"),
            updatedAtMillis = parseTimestampMillis(market.stringOrNull("updated_time")),
            openTimeMillis = parseTimestampMillis(market.stringOrNull("open_time")),
            closeTimeMillis = parseTimestampMillis(market.stringOrNull("close_time")),
            imageUrl = imageUrl,
            webUrl = webUrl,
        )
    }

    private fun extractImageUrl(event: JsonObject, market: JsonObject): String? {
        // Check product_metadata.image in event then market
        for (obj in listOf(event, market)) {
            val metadata = obj["product_metadata"]?.jsonObject ?: continue
            metadata["image"]?.let { imageVal ->
                val str = when (imageVal) {
                    is JsonPrimitive -> imageVal.contentOrNull
                    else -> null
                }
                if (!str.isNullOrEmpty()) return str
            }
        }
        return null
    }

    private fun normalizeUnitPrice(value: Double?): Double? {
        if (value == null) return null
        return if (value > 1.0) value / 100.0 else value
    }

    private fun parseTimestampMillis(isoString: String?): Long? {
        if (isoString.isNullOrEmpty()) return null
        return try {
            java.time.Instant.parse(isoString).toEpochMilli()
        } catch (_: Exception) {
            try {
                // Try without fractional seconds
                java.time.Instant.parse(isoString.replace(Regex("\\.\\d+"), "")).toEpochMilli()
            } catch (_: Exception) {
                null
            }
        }
    }
}

// --- JSON helper extensions ---

/** Read a string field, tolerating missing/null. */
internal fun JsonObject.stringOrNull(key: String): String? {
    val element = this[key] ?: return null
    return if (element is JsonPrimitive) element.contentOrNull else null
}

/**
 * Flexibly read a number that might be encoded as int, float, or string.
 */
internal fun JsonObject.flexibleDouble(key: String): Double? {
    val element = this[key] ?: return null
    if (element is JsonNull) return null
    if (element is JsonPrimitive) {
        element.doubleOrNull?.let { return it }
        element.longOrNull?.let { return it.toDouble() }
        element.contentOrNull?.toDoubleOrNull()?.let { return it }
    }
    return null
}

internal fun JsonObject.flexibleLong(key: String): Long? {
    val element = this[key] ?: return null
    if (element is JsonNull) return null
    if (element is JsonPrimitive) {
        element.longOrNull?.let { return it }
        element.intOrNull?.let { return it.toLong() }
        element.contentOrNull?.toLongOrNull()?.let { return it }
    }
    return null
}
